prep_ubuntu.sh

```diff
@@ -24,7 +24,7 @@ upgrade_base_operating_system() {
   cd /opt/scripts/tools/
   git pull
   # Update to the latest ti kernel with initrd!
-  FORCEMACHINE=TI_AM335x_BeagleBoneBlack sh update_kernel.sh --lts-4_14 --ti-kernel
+  FORCEMACHINE=TI_AM335x_BeagleBoneBlack sh update_kernel.sh --lts-4_14 --bone-kernel
```

Comment:
Does this choice affect the PRU (or other) code in Redeem?

Reply:
I'm guessing not, or @ThatWileyGuy wouldn't have done it. Have you tested the image generated yet Andrew?

Reply:
Then I'm confused about the different kinds of kernels. There are the "mainstream" kernels (like the ones Ubuntu installs). I think they lack any interface that we need for the PRUs. Then @ThatWileyGuy had things running on a "bone" kernel (from Nelson's repo) that provided one form of interface. But he indicated that he wanted to change to a different interface because it provided a "better" way for us to communicate with the PRUs. I thought that resulted in changes to the code generation for the PRUs. So, I see him now changing back to a "bone" kernel. Therefore, I expected some change in Redeem to match it.

Reply:
From my understanding it's actually the way the overlays were loaded and initialized on top of the kernel that made us choose one over the other. Also one kernel allows only 1 comm method, while iirc the other allows both.

Reply:
So, back to my original question about changing the kernel, I'm guessing the @ThatWileyGuy prior change in kernel was only in preparation for a change to Redeem which has yet to be offered.

Reply:
Yes, I have flashed the generated image onto a test board and verified that I get motor movement.

There are two ways to communicate with the PRU today:
- UIO exposes the PRU's memory and config pages to userspace, where an application asks the kernel to map them into its memory space. It then loads firmware and manipulates the PRUs directly and uses syscalls to wait for interrupts the PRUs send the main CPU.
- RPROC initializes PRU firmware in a kernel driver and parses special tables in the firmware to set up communications channels in /dev. Userspace processes can then read() and write() to the nodes in /dev to communicate with the running firmware via a message API. All interrupts from the PRU to the CPU are used to indicate that messages are being sent and are thus handled inside the kernel driver doing the message passing.

The TI kernels can do both UIO and RPROC. The Bone kernels can only do UIO. Redeem currently uses UIO.

I've been hoping to move us over to RPROC at some point because it makes it much easier to add a lot more sophistication to the PRU firmware, which I believe will help with things like baby stepping and other features that need PRU behavior that's more complicated than carrying out a long list of steps from a queue.

The PRU programming community is quite polarized on this subject - there's a perception that TI is trying to force everyone to switch to RPROC because it's easier to maintain (RPROC is a generic Linux framework for interacting with coprocessors, whereas UIO seems to be mostly tied to the Beaglebone). There's also a perception that RPROC is a higher-overhead framework and that it's silly because the message-passing facilities RPROC provides can be built within UIO.